IPNI_FEDERATION.md

```diff
@@ -0,0 +1,425 @@
+  link to the previous snapshot. IPNI instances can share snapshots with each other to quickly update and reconcile
+  their content.
+
+## Specification
+
+The IPNI federation protocol is designed to address the challenges that arise when multiple IPNI instances operate
+concurrently across different regions or domains. The fundamental premise of this specification is to provide a
+systematic method for these instances to interact, synchronize, and offer a unified view of the index records to
+end-users. To achieve this, each node in a federated IPNI network performs the following:
+
+1. Initialization - the process by which a node joins a federated network.
+2. Periodic snapshots - the process by which a node periodically captures an immutable snapshot of its providers along
+   with associated metadata.
+3. Snapshot exchange - the process by which each node propagates their latest snapshot to the rest of the network.
+4. Reconciliation - the process by which snapshots received from other nodes is reflected on to bring the network state
+   knowledge up-to-date.
+
+Finally, the specification expands on cases where conflict might occur and how a node should resolve it.
+
+### Initialization
+
+At the outset, every participating IPNI instance undergoes an initialization process. This involves setting up its
+vector clock to the default or baseline state and establishing secure communication channels with other participating
+instances. It's during this phase that instances also share their initial state and any configuration parameters crucial
+for the federation protocol. The initialization ensures a foundational layer of consistency, preparing the system for
+subsequent operations like snapshot exchange and reconciliation.
+
+### Periodic Snapshot
+
+Recognizing that continuous real-time synchronization could be resource-intensive and inefficient, the IPNI federation
+protocol employs a periodic snapshot mechanism. At predetermined intervals, or upon detecting significant changes, each
+IPNI instance captures a comprehensive snapshot of its index. This snapshot, which is essentially a compact
+representation of the instance's current state, serves as the primary data unit for synchronization among instances. It
+ensures that instances can exchange substantial amounts of data without overwhelming the network or causing excessive
+data traffic.
+
+The snapshot mechanism also offers a strategic advantage. In the event of any system failures or communication
+breakdowns, snapshots can be used to quickly restore or reconcile the state of an instance, minimizing data loss and
+ensuring rapid recovery.
+
+A snapshot consists of:
+
+* **Epoch**: a monotonically increasing value that represents the time at which snapshot was formed.
+* **Vector Clock**: the monotonically increasing vector clock value that corresponds to the IPNI instance.
+* **Provider to Ingest State Map**: a map of provider ID to the latest advertisement CID processed by the IPNI
```

Comment:
Does "processed" mean there's an expectation here that if a client see "height 10" that all of the advertisements up to height 10 will be queryable (by clients or other IPNI nodes), or does that mean they've simply heard about the latest CID/height. Similarly, must height have been validated?

Note: I get that you may not want to be prescriptive here around caching / loadbalancing strategies (e.g. could a client end up with inconsistent state due to a load balancer routing the data to different backends rather than being sticky), either way it'd probably be helpful to call this out for any client implementers.